### PR TITLE
[Layers] Add SpatialDropout support and enhance Dropout

### PR DESCRIPTION
**Pull request by jon-tow**
# Summary
### Adds:
- Support for `SpatialDropout1D`, `SpatialDropout2D`, and `SpatialDropout3D`.
- `Dropout` testing for forward-pass computation.
### Todo:
- Test derivatives for each dropout layer.
### Related Issue: 
#54

### CODE REVIEW

File: Sources/TensorFlow/Operators/NN.swift [hunk 1]
```diff
@@ -12,6 +12,44 @@
...
+///
+/// - Parameters:
+///     - input: The input units.
+///     - probability: The probability that a input unit will be dropped.
+///     - noiseShape: The shape of the randomly generated keep mask.
+/// - Precondition: `probability` is within the range 0..<1.
+/// - Note: `dropout` is an identity function on `input` when called during the inference phase of
+///     learning.
+@differentiable(wrt: input)
+public func dropout<Scalar: TensorFlowFloatingPoint>(
```
Comment (eaplatanios) [inline Sources/TensorFlow/Operators/NN.swift]: Why did you also add a global function aside from `Tensor.droppingOut`? I believe we should not be mixing `Context`s in the operator definitions and stick to only using them within `Layer`s so we don't mix up high-level and low-level APIs. @rxwei what do you think?

File: Sources/TensorFlow/Operators/NN.swift [hunk 2]
```diff
@@ -12,6 +12,44 @@
...
+
+/// Returns the result of applying dropout on an input tensor.
+///
+/// - Parameters:
+///     - input: The input units.
+///     - probability: The probability that a input unit will be dropped.
+///     - noiseShape: The shape of the randomly generated keep mask.
+/// - Precondition: `probability` is within the range 0..<1.
+/// - Note: `dropout` is an identity function on `input` when called during the inference phase of
+///     learning.
```
Comment (eaplatanios) [inline Sources/TensorFlow/Operators/NN.swift]:
```suggestion
///   learning.
```

File: Sources/TensorFlow/Operators/NN.swift [hunk 3]
```diff
@@ -12,6 +12,44 @@
...
 // limitations under the License.
 
+//===------------------------------------------------------------------------------------------===//
+// Core
+//===------------------------------------------------------------------------------------------===//
+
+private extension Tensor where Scalar: TensorFlowFloatingPoint {
+    /// Returns a dropout given a probability and noise shape.
+    @differentiable(wrt: self where Scalar: Differentiable)
+    func droppingOut(probability: Double, noiseShape: TensorShape) -> Tensor {
```
Comment (eaplatanios) [inline Sources/TensorFlow/Operators/NN.swift]: Why not make this public?

Comment (eaplatanios) [inline Sources/TensorFlow/Operators/NN.swift]:
```suggestion
    func droppingOut(withProbability probability: Double, noiseShape: TensorShape) -> Tensor {
```

File: Sources/TensorFlow/Operators/NN.swift (hunk 1 above)
Comment (jon-tow) [reply Sources/TensorFlow/Operators/NN.swift]: As you noted below, I restricted `Tensor.droppingOut` to `private` because it doesn't make sense, at least for me, to expose sort of half an implementation of `dropout` given that the scaling of samples in `Tensor.droppingOut` is useful only in the context of learning phases. I don't find `Tensor.droppingOut`, defined as inverted dropout, to be useful to the outside world given its purpose.

Now the purpose of abstracting out the `dropout` function with `Context` was to avoid repetitively switching on `learningPhase` for each of the four dropout layers' `callAsFunction`. It was just a reduction of ~24 lines. But I have no problem bringing that back into the layer and providing `Tensor.droppingOut` publicly if that's desired.

File: Sources/TensorFlow/Operators/NN.swift (hunk 3 above)
Comment (rxwei) [reply Sources/TensorFlow/Operators/NN.swift]: “With” is often considered as a meaningless word in argument labels in Swift APIs (though it’s not formally documented). We’ve been avoiding “with” unless it’s both semantically and grammatically significant.

File: Sources/TensorFlow/Layers/Core.swift [hunk 4]
```diff
@@ -12,54 +12,150 @@
...
     @noDerivative public let probability: Double
 
     /// Creates a dropout layer.
     ///
-    /// - Parameter probability: The drop probability.
+    /// - Parameter probability: The probability that an input unit will be dropped. This value must
+    ///   be between `0` and `1`, i.e. `(0..<1)`.
     public init(probability: Double) {
+        precondition(
+            (0..<1.0).contains(probability),
```
[the remark this answers is not in the document]
Comment (rxwei) [reply Sources/TensorFlow/Layers/Core.swift]: Yep.

File: Sources/TensorFlow/Operators/NN.swift (hunk 1 above)
Comment (jon-tow) [reply Sources/TensorFlow/Operators/NN.swift]: @eaplatanios, @rxwei: 
I'm going to upgrade `Tensor.droppingOut(probability:noiseShape:)` to public in `NN.swift`. 

Should I move `dropout(_:probability:noiseShape:)` to `Layers/Core.swift` as a free function with private access or move its contents, the switch on learning phase, into each of the dropouts' `callAsFunction` just as it was previously handled in `Dropout`? 
Thanks!

File: Sources/TensorFlow/Layers/Core.swift [hunk 5]
```diff
@@ -12,54 +12,166 @@
...
+
+    /// Creates a 1-D spatial dropout layer.
+    ///
+    /// - Parameter probability: The probability that an input unit will be dropped. This value must
+    ///   be in the interval `0..<1`.
+    public init(probability: Double) {
+        precondition(
+            (0..<1.0).contains(probability),
+            "Dropout probability must be between 0 and 1. Got: \(probability)")
+
```
Comment (rxwei) [inline Sources/TensorFlow/Layers/Core.swift]: Remove redundant empty line.

File: Sources/TensorFlow/Layers/Core.swift [hunk 6]
```diff
@@ -12,54 +12,166 @@
...
+    /// Returns the output obtained from applying the layer to the given input.
+    ///
+    /// - Parameter input: The input to the layer of shape
+    ///   [batch count, depth, height, width, channel count]
+    /// - Returns: An output of the same shape as `input`.
+    /// - Precondition: `input` must be rank `5`.
+    @differentiable
+    public func callAsFunction(_ input: Tensor<Scalar>) -> Tensor<Scalar> {
+        dropout(input, probability: probability,
+             noiseShape: [input.shape[0], 1, 1, 1, input.shape[4]])
```
Comment (rxwei) [inline Sources/TensorFlow/Layers/Core.swift]: Fix indentation.

File: Sources/TensorFlow/Layers/Core.swift [hunk 7]
```diff
@@ -12,54 +12,166 @@
...
+
+    /// Creates a 3-D spatial dropout layer.
+    ///
+    /// - Parameter probability: The probability that an input unit will be dropped. This value must
+    ///   be in the interval `0..<1`.
+    public init(probability: Double) {
+        precondition(
+            (0..<1.0).contains(probability),
+            "Dropout probability must be between 0 and 1. Got: \(probability)")
+
```
Comment (rxwei) [inline Sources/TensorFlow/Layers/Core.swift]: Remove redundant empty line.

File: Tests/TensorFlowTests/LayerTests.swift [hunk 8]
```diff
@@ -361,6 +361,20 @@ final class LayerTests: XCTestCase {
...
     }
 
+    func testDropout() {
+        Context.local.learningPhase = .training
+        let probability = 0.25
+        let layer = Dropout<Double>(probability: probability)
+        let input = Tensor<Double>(randomNormal: [500, 500], mean: Tensor(5))
+        let output = layer(input)
+        let fractionDropped = Tensor<Double>(output .== 0).mean()
+        XCTAssertTrue(fractionDropped.isAlmostEqual(to: Tensor(probability), tolerance: 0.01))
```
Comment (rxwei) [inline Tests/TensorFlowTests/LayerTests.swift]:
```suggestion
        XCTAssertEqual(fractionDropped.scalar!, probability, tolerance: 0.01)
```
I forgot the exact argument label but `XCTAssertEqual` has an approximating variant.

File: Tests/TensorFlowTests/LayerTests.swift [hunk 9]
```diff
@@ -361,6 +361,20 @@ final class LayerTests: XCTestCase {
         XCTAssertEqual(output, expected)
     }
 
+    func testDropout() {
```
Comment (rxwei) [inline Tests/TensorFlowTests/LayerTests.swift]: I'd suggest using `withLearningPhase(_:_:)` in tests to make scoping clear and prevent global effects.

**Comment by jon-tow** [comment]
@rxwei @eaplatanios 
Closing this PR. `SpatialDropout` layers seem too narrow and may be better suited in a higher abstraction level API (e.g. SwiftAi). I might make a PR for `droppingOut(probability:noiseShape:)` since it lends nicely to spatial dropout generalizations.
